stdlib/mexpr/parser.mc

```diff
@@ -0,0 +1,448 @@
+      eatWSAC p str
+    in remove p xs
+end
+
+-- Eat multiline comment of the form *-  -*
+lang MultilineCommentParser = WSACParser
+  sem eatWSAC (p : Pos) =
+  | "/-" ++ xs ->
+    recursive
+    let remove = lam p. lam str. lam d.
+      match str with "/-" ++ xs then remove (advanceCol p 2) xs (addi d 1) else
+      match str with "\n" ++ xs then remove (advanceRow p 1) xs d else
+      match str with "-/" ++ xs then
+        if eqi d 1 then eatWSAC (advanceCol p 2) xs
+        else remove (advanceCol p 2) xs (subi d 1) else
+      match str with [_] ++ xs then remove (advanceCol p 1) xs d else
+      if eqi d 0 then eatWSAC p str else posErrorExit p "Unmatched multiline comments."
+    in remove (advanceCol p 2) xs 1
+end
+
+-- Commbined WSAC parser for MExpr
+lang MExprWSACParser = WhitespaceParser + LineCommentParser + MultilineCommentParser
+
+let _ = use MExprWSACParser in
+  utest eatWSAC (initPos "") " --foo \n  bar "
+    with {str = "bar ", pos = posVal "" 2 2} in
+  utest eatWSAC (initPos "") " /- foo -/ bar"
+    with {str = "bar", pos = posVal "" 1 11} in
+  utest eatWSAC (initPos "") " /- foo\n x \n -/ \nbar "
+    with {str = "bar ", pos = posVal "" 4 0} in
+  utest eatWSAC (initPos "") " /- x -- y /- foo \n -/ -/ !"
+    with {str = "!", pos = posVal "" 2 7} in
+  ()
+
+
+
+-- Top of the expression parser. Connects WSAC with parsing of other non terminals
+lang ExprParser = WSACParser
+  sem parseExpr (p: Pos) =
+  | s ->
+    let r1 = parseExprMain p 0 s in
+    let r2 = eatWSAC r1.pos r1.str in
+    if eqi (length r2.str) 0 then r1.val
+    else posErrorExit r2.pos "Parse error. Unknown characters."
+  
+  sem parseExprMain (p: Pos) (prec: Int) =
+  | s ->
+    let r1 = eatWSAC p s in
+    let exp = parseExprImp r1.pos r1.str in
+    let r2 = eatWSAC exp.pos exp.str in
+    parseInfix r2.pos prec exp r2.str 
+
+  sem parseInfix (p: Pos) (prec: Int) (exp: Expr) =
+
+  sem parseExprImp (p: Pos) =
+  | _ -> posErrorExit p "Parse error. Unknown character sequence."
+end
+
+
+-- Include this fragment if there are no infix operations
+lang ExprParserNoInfix = ExprParser
+  sem parseInfix (p: Pos) (prec: Int) (exp: Expr) =
+  | _ -> exp
+end
+
+-- Parses an identfier that starts with a lower-case letter or a '_' if parameter
+-- 'upper' is false, and starts with an upper-case letter if 'upper' is true.
+-- The rest of the string can contain both upper and lower-case letters.
+-- If no identifier, the 'val' field contains an empty string.
+let parseIdent = lam upper. lam p. lam str.
+  recursive
+  let work = lam acc. lam first. lam p. lam str.
+    match str with [x] ++ xs then
+      let c = char2int x in
+      let m1 = or (not first) upper in
+      let m2 = or (not first) (not upper) in
+      if or (or (and m1 (and (geqi c 65) (leqi c 90)))
+                (and m2 (or (eqi c 95) (and (geqi c 97) (leqi c 122)))))
+	    (and (not first) (and (geqi c 48) (leqi c 57)))	    
```

Comment:
I am not so keen on having these magic numbers directly in the code. Would it perhaps be better to use something like `leqChar` and `geqChar`? In that case you actually see the characters directly, so that you do not have to keep looking up the ASCII table when debugging. Though I would prefer to have something like `isUpper`, `isLower`, and `isDigit`.

Reply:
OK, I fixed it now and used the functions in char.mc.